src/machines/SyncMachine.ts

```diff
@@ -0,0 +1,307 @@
+        },
+        [CONNECTED_STATES.handlingUnhandledEvent]: {
+          id: CONNECTED_STATES.handlingUnhandledEvent,
+          invoke: {
+            src: 'updateLastSyncedEvent',
+            onDone: {
+              actions: ['sendAck'],
+              target: 'idle',
+            },
+            onError: `#${SYNC_MACHINE_STATES.ERROR}`,
+          },
+        },
+        [CONNECTED_STATES.handlingMetadataChanged]: {
+          id: 'handlingMetadataChanged',
+          initial: 'detectingDiff',
+          states: {
+            detectingDiff: {
+              invoke: {
+                src: 'metadataDiff',
+                onDone: { actions: ['metadataDecided'] },
+              },
+              on: {
+                METADATA_DECIDED: [
+                  { target: `#${CONNECTED_STATES.handlingVoidedTx}`, cond: 'metadataVoided', actions: ['unwrapEvent'] },
+                  { target: `#${CONNECTED_STATES.handlingUnvoidedTx}`, cond: 'metadataUnvoided', actions: ['unwrapEvent'] },
+                  { target: `#${CONNECTED_STATES.handlingVertexAccepted}`, cond: 'metadataNewTx', actions: ['unwrapEvent'] },
+                  { target: `#${CONNECTED_STATES.handlingFirstBlock}`, cond: 'metadataFirstBlock', actions: ['unwrapEvent'] },
+                  { target: `#${CONNECTED_STATES.handlingUnhandledEvent}`, cond: 'metadataIgnore' },
+                ],
+              },
+            },
+          },
+        },
+        // We have the unchanged guard, so it's guaranteed that this is a new tx
+        [CONNECTED_STATES.handlingVertexAccepted]: {
+          id: CONNECTED_STATES.handlingVertexAccepted,
+          invoke: {
+            src: 'handleVertexAccepted',
+            data: (_context: Context, event: Event) => event,
+            onDone: {
+              target: 'idle',
+              actions: ['sendAck', 'storeEvent', 'updateCache'],
+            },
+            onError: `#${SYNC_MACHINE_STATES.ERROR}`,
+          },
+        },
+        [CONNECTED_STATES.handlingVoidedTx]: {
+          id: CONNECTED_STATES.handlingVoidedTx,
+          invoke: {
+            src: 'handleVoidedTx',
+            data: (_context: Context, event: Event) => event,
+            onDone: {
+              target: 'idle',
+              actions: ['storeEvent', 'sendAck', 'updateCache'],
+            },
+            onError: `#${SYNC_MACHINE_STATES.ERROR}`,
+          },
+        },
+        [CONNECTED_STATES.handlingUnvoidedTx]: {
+          id: CONNECTED_STATES.handlingUnvoidedTx,
+          invoke: {
+            src: 'handleUnvoidedTx',
+            data: (_context: Context, event: Event) => event,
+            onDone: {
+              // The handleUnvoidedTx will remove the tx from the database, we should
+              // re-add it:
+              target: `#${CONNECTED_STATES.handlingVertexAccepted}`,
+              // We shouldn't send ACK, as we'll send the ACK after handlingVertexAccepted
+              actions: ['storeEvent', 'updateCache'],
```

Comment:
Should we have a new test to cover this recently found bug?

Reply:
I will create a integration test that cover this and the other bugs we found in another PR, after the monorepo PR is merged

Reply:
Could you create a KTLO issue for this and link it here, please? So we can keep track of it

Reply:
Issue: https://github.com/HathorNetwork/hathor-wallet-service-sync_daemon/issues/95

Reply:
Yes

Issue: https://github.com/HathorNetwork/hathor-wallet-service-sync_daemon/issues/96

Reply:
Thanks!